### DISCUSSION TITLE
Top level dynamic route /[[...node]]/page.js shallow routing causes re-render

### DISCUSSION THREAD

**Discussion by knowncolor** [vercel/next.js#55534] [App Router]
I have a single page, `app/[[...node]]/page.js` which should show details about the current node.
I would like to update the browser url at the top-level as the use clicks around my app, without re-rendering the page.

`localhost:3000/node1`,  `localhost:3000/node2`, etc

When I do something like `router.push(startNode.name, undefined, { shallow: true });` it always forces a re-rendering of the page.

Anyway to avoid?

If I switch it to instead push as a url param `router.push("?node=" + startNode.name, undefined, { shallow: true });` then it works.
The [caveat in the docs](https://nextjs.org/docs/pages/building-your-application/routing/linking-and-navigating#caveats) states "Shallow routing only works for URL changes in the current page." but to my understanding I am staying on the same page (I only have one page).

Any help appreciated!

**Comment by DIEGOHORVATTI** [comment]
> I have a single page, `app/[[...node]]/page.js` which should show details about the current node. I would like to update the browser url at the top-level as the use clicks around my app, without re-rendering the page.
> 
> `localhost:3000/node1`, `localhost:3000/node2`, etc
> 
> When I do something like `router.push(startNode.name, undefined, { shallow: true });` it always forces a re-rendering of the page.
> 
> Anyway to avoid?
> 
> If I switch it to instead push as a url param `router.push("?node=" + startNode.name, undefined, { shallow: true });` then it works. The [caveat in the docs](https://nextjs.org/docs/pages/building-your-application/routing/linking-and-navigating#caveats) states "Shallow routing only works for URL changes in the current page." but to my understanding I am staying on the same page (I only have one page).
> 
> Any help appreciated!

 Did you try to use a layout involving the pages you want to click on and use NextLink instead of push?

**Reply by knowncolor** [reply]
Hi Diego - thanks for the suggestion!

My page is pretty much just a fullscreen [reagraph](https://reagraph.dev/?path=%2Fstory%2Fdemos-nodes--icons) which provides me click events - so I need to use the `router/push` to update the url `localhost:3000/node` as the user clicks different nodes.

Perhaps the entire thing should be in layout?

**Reply by DIEGOHORVATTI** [reply]
Yes, this way it works better, at least that's the way I found it to be done, but nothing prevents you from using replace. I don't think this would update the entire component, but it still has to be in a layout, the layout that stores the state referring to each route


![image](https://github.com/vercel/next.js/assets/72998710/98800623-4b13-47ba-af8b-e41931154605)
